### ISSUE TITLE
Improve `printf` behavior with multiple arrays as arguments

### ISSUE THREAD

**Issue by mqudsi** [fish-shell/fish-shell#5836]
When you first see this, you think "this is cool!" and wonder how you can simplify your code with the power of cartesian expansion:

```fish
> printf '"%s", ' (echo one\ntwo\three)
"one", "two", "three", ¶  
```

then you see this and you become sad again:

```fish
> printf '"%s": "%s",\n' (echo one\ntwo\nthree) (echo a\nb\nc)
"one": "two", 
"three": "a", 
"b": "c",
```

While technically `printf` (as a builtin) is privy to additional information about parameters such that it could *theoretically* tell the difference between `printf ... (echo foo; echo bar) (echo baz; echo qux)` and `printf foo bar baz qux`, there's no precedent for that in the codebase.

(I would have written everything with braces (e.g. `printf {foo, bar} {baz, qux}`) which I believe can  also be expanded to two arrays rather than 4 variables but this is clearer).

But fortunately it isn't necessary to grapple with enriching fish's variable expansion / argument passing semantics to improve this. If we keep the same "expand all arrays into arguments and then invoke the command with the flattened result" behavior, we can still get this behavior:

```fish
> printf '"%s": "%s",\n' (echo one\ntwo\nthree) (echo a\nb\nc)
"one": "a",
"two": "b",
"three": "c",
```

by simply counting the number of `%` placeholders in advance, and divide that by the number of arguments to determine the parameters to set aside for each variable.

I honestly cannot see anywhere where the current behavior is desired (since extra parameters are not dropped), whereas the proposed behavior would allow certain uses of `for` loops to be replaced with simple cartesian expansion.

**Comment by 0ion9** [comment]
This seems kind of like an argument for the functional programming concept `zip`, honestly.
Here's a basic implementation which could be optimized a lot:

```
function _zip -d "zip <array names> interleaves the elements of the arrays"
  set i 1
  while true
    # zip stops once you run out of 'whole sets'
    # (ie. not all of the arrays have an element set in the given position)
    for arrname in $argv
       set arrname $arrname"[$i]"
       if not set -q $arrname
         return
       end
    end
    for arrname in $argv
       set arrname $arrname"[$i]"
       printf %s\n $$arrname
    end
    set i (math $i+1)
  end
end
```

The main problem with the above is it doesn't support slicing (eg `_zip a[2..-1] b[2..-1]`)

Personally, I favor some kind of explicit construct like zip, because I think it's confusing to the user for $foo $bar to imply a different sequencing of arguments only for the `printf` builtin -- that seems more bashy than fishy IMO.

**Comment by mqudsi** [comment]
@0ion9 indeed, I was trying to avoid directly bringing up zip as I was focusing strictly on a printf intuitiveness "bug".

> Personally, I favor some kind of explicit construct like zip, because I think it's confusing to the user for $foo $bar to imply a different sequencing of arguments only for the printf builtin -- that seems more bashy than fishy IMO.

I see what you are saying, but I think that's only the case if you're strictly looking at it from an implementation point-of-view, taking into account that the arrays in `printf $arr1 $arr2` are first expanded before being passed to `printf` as arguments, rather than the arrays themselves being the arguments to `printf`. I *think* that for the typical fish user that might not be the expectation.

And if you *aren't* using arrays (or process substitution) then there's basically no case where you would be passing an indefinite number of arguments to a `printf` call with a fixed number of `%` placeholders, no?

**Comment by 0ion9** [comment]
> I see what you are saying, but I think that's only the case if you're strictly looking at it from an implementation point-of-view, taking into account that the arrays in printf $arr1 $arr2 are first expanded before being passed to printf as arguments, rather than the arrays themselves being the arguments to printf. 

The user is required to deal with the reality that commands receive arguments as a single array of strings. That's not an implementation detail of fish.  For example, supposing that the user invokes `printf` (with the revision you propose applied) like this:

`printf '%d\t%s\n' $ranks $filenames

with ranks =`1 2 3 4`
and filenames = `foo.png bar.png`

That provides 3 pairs:

* 1 4
* 2 foo.png
* 3 bar.png

and the fact that printf is receiving a single list of arguments becomes apparent (the output '1 4', with 2 elements from the same array, could not occur otherwise).

This is IMO a perfectly plausible scenario as soon as the arrays are being generated by script.

**Comment by faho** [comment]
>I don't know if the latter has any prevalent usages in the fish codebase/userbase.

Checking the output of `grep "printf.*%.*%" -r share/`, nothing springs up.

But that doesn't change that `printf` is fundamentally the wrong place to address this.

Instead, how about, yes, a `zip` builtin, that you'd use like

```fish
printf '"%s": "%s",\n' (zip 2 (echo one\ntwo\nthree) (echo a\nb\nc))
```

zip would split its arguments into groups, and then print

```
one
a
two
b
three
c
```

The obvious drawback being that it loses information, so

1. The number of groups is required - you need `zip 2 $foo $bar`
2. If $foo and $bar have an unequal number of elements, this indeed breaks (should it just error out if the number of args isn't divisible by the number of groups?). The alternative is custom parsing since e.g. any delimiter could also appear in the input.

Also:

- The name is a problem. `zip` is kind of a well-known archival tool. One possiblity is to change the name, another to put it under the `string` builtin as `string zip`.

**Comment by 0ion9** [comment]
You could also consider this operation as Transpose - transposing the dimensions of an array (in which each row/scanline is one actual fish array). You put in data in row order (l->r,t->b), and get it out in column order (t->b, l->r).

Another point: using a separator argument would dispense with the problems of error checking. As in `zip \; $numbers \; $names` -- the first instance of \; specifies the separator, subsequent instances mean 'end of this array'
(I'm personally tempted to use `--` but acknowledge that would probably be confusing.)

**Comment by faho** [comment]
>Another point: using a separator argument would dispense with the unequal-length arrays

Yeah, but... we can't! Any separator could appear in the arguments themselves!

Any `--` can appear in a variable, even if people will generally avoid that.

**Comment by 0ion9** [comment]
... that was the purpose of, in my example, obligating the user to specify the separator string (first argument). Is it really implausible that the user can determine what string will not be present?

**Comment by faho** [comment]
>Is it really implausible that the user can determine what string will not be present in their particular case?

Yes. What if you're not using string literals, but some commands output? Do you really want to have to check if it includes a tab (plenty of commands do)?

Also it's annoying and error-prone.

**Comment by 0ion9** [comment]
.. What commands output a single tab on it's own line? I use tsv type data, cut, awk, etc all the time, and I can't think of any. 

Your last point is certainly true though. I suppose it's better to force the user to check the array lengths beforehand.
